src/server/master_server/chunk_server_manager.h

```diff
@@ -54,45 +137,132 @@ class ChunkServerManager {
+  // This returns all the chunk server locations where this chunk is stored.
+  // This returns an empty set if there is no location. This can be because the
+  // chunk hasn't been allocated by calling the AllocateChunkServer (dev bug) or
+  // because the chunk servers where this chunk was allocated have been
+  // unregistered (due to server down). The returned locations might be smaller
+  // than the number of allocated servers if some of those servers have been
+  // unregistered (due to server down). The ChunkServer heartbeat monitor,
+  // checks when a server is down and unregister it. If the server comes back
+  // up, it will be reregistered by the chunkservermanager service.
+  ChunkServerLocationThreadSafeFlatSet GetChunkLocations(
+      const std::string& chunk_handle);
 
  private:
-  // TODO(bmokutub): Consider a thread safe map or locking this.
-  // May not need to use shared_ptr since proto manage the memory?
-  absl::flat_hash_map<protos::ChunkServerLocation,
-                      std::shared_ptr<protos::ChunkServer>,
-                      ChunkServerLocationHash>
+  // This is a thread safe map that stores the mapping from chunkserverlocation
+  // to the chunkserver. The location uniquely identifies the chunkserver. This
+  // is used to store the registered chunk servers and are removed when the
+  // chunkserver is unregistered.
+  gfs::common::thread_safe_flat_hash_map<protos::ChunkServerLocation,
+                                         std::shared_ptr<protos::ChunkServer>,
+                                         ChunkServerLocationHash>
       chunk_servers_map_;
 
+  // This is a thread safe map that stores the mapping from chunk handle to the
+  // chunk server locations where it is stored. Chunk handles are added here
+  // once chunkservers are allocated for them. Using a thread safe set since we
+  // concurrently add and remove chunk locations for a chunk handle.
+  gfs::common::thread_safe_flat_hash_map<std::string,
+                                         ChunkServerLocationThreadSafeFlatSet>
+      chunk_locations_map_;
+
+  // This is used to keep a priority list of chunk servers for chunk allocation.
+  // Servers with most available disk are given highest priority. The reason for
+  // using a list and not a heap is because we need to be able to remove a chunk
+  // server from the list whenever a chunkserver is unregistered/down. Heaps
+  // don't support random removal. We don't want to be allocating chunks to an
+  // unavailable chunk server. And this also helps us to do load balancing, by
+  // prioritizing chunk servers with most available disk first for chunk
+  // allocation. we always keep this list sorted so max available disk
+  // chunkservers first. There are several possible optimizations for this, but
+  // will fist monitor performance of this.
+  std::list<std::shared_ptr<protos::ChunkServer>> chunk_servers_priority_list_;
```

Comment:
Based on the impl shown, it seems that this should be an ordered structure, either an std::set, or some equivalent structure in absl::container.

However, as I mentioned above, maybe we can just simplify things and use randomized algorithm to select chunk servers, that way, this can just be a vector. 

Btw, why a list not a vector?

Reply:
We can't use an ordered set here because we are ordering by disk only. And using an ordered set will prevent two chunkservers from having the same disk space, the order is also used for equality check. Two objects a and b are considered equivalent if neither compares less than the other: !comp(a, b) && !comp(b, a).

Forgot to also add this in the comment. I have responded to your randomized algo comment above. Also a list is more efficient here for the use case than a vector.

Reply:
It sounds like you should consider std::multiset<>, which allows multiple equivalent items to exist. Mind taking a look? The reason I ask is that I think the manual sorting could have been avoided everything when something changed by using a set / multiset.

Reply:
Multiset also does this: Two objects a and b are considered equivalent if neither compares less than the other: !comp(a, b) && !comp(b, a).

Reply:
Uses the comparision for equality and thats why I didn't use it. If i say erase ChunkLocation a (that has disk 20), it will remove all chunk locations with disk size 20.

Reply:
I can avoid that by manually iterating to delete and compare the chunkservers. Also when a chunkserver disk is updated, you have to remove the chunkserver from the set and readd it so it can be placed in the right location. So still not a perfect structure for this use case. And this doesn't remove the sorting work, because it still does sorting for us instead of we doing the sorting. One benefit I see of we externalizing the sorting is that it can be done in the background since the number of chunkservers can be large (thousands.)

Reply:
Okay since this is a protoype, will use a multiset, but as mentioned this will always do sorting in foreground. But since in this prototype we most likely wouldn't have thousands of node, may be okay.

Reply:
On second thought, leaving this like this. The multiset doesn't give us much benefit and I still need to do some patch work to make it support this scenario fully. Infact I will still need to call sort on the set explicitly. I will stay with the list implementation as it simplifies some operation and allows background sorting or periodic sorting. Happy to chat more and explain this on slack.

Reply:
I am fine with leaving this as is, as we have a lot other priorities to catch up.